rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,686 @@
+
+Although these restrictions require significant changes, it does not prevent
+known upcoming migrations such as [sharing saved-objects in multiple spaces](https://github.com/elastic/kibana/issues/27004) or [splitting a saved
+object into multiple child
+documents](https://github.com/elastic/kibana/issues/26602). To ensure that
+these migrations are idempotent, they will have to generate new saved object
+id's deterministically with e.g. UUIDv5.
+
+
+### 4.2.1.2 Migration algorithm: Cloned index per version
+Note:
+- The description below assumes the migration algorithm is released in 7.10.0.
+  So < 7.10.0 will use `.kibana` and >= 7.10.0 will use `.kibana_current`.
+- We refer to the alias and index that outdated nodes use as the source alias
+  and source index.
+- Every version performs a migration even if mappings or documents aren't outdated.
+
+1. Locate the source index by fetching aliases (including `.kibana` for versions prior to v7.10.0) `GET '/_alias/.kibana_current,.kibana_7.10.0,.kibana'`. The source index is the index the `.kibana_current` alias points to, or if it doesn’t exist, the index the `.kibana` alias points to.
+2. If `.kibana_current` and `.kibana_7.10.0` both exists and are pointing to the same index this version's migration has already been completed.
+   1. Because the same version can have plugins enabled at any point in time, perform the mappings update in step (6).
+   2. If the target index contains any documents with an outdated migrationVersion number, log an error explaining that a re-migration is required and mark the saved object type these documents belong to as unavailable. See section (4.2.1.4).
+   3. Skip to step (9) to start serving traffic.
+3. If `.kibana_current` is pointing to an index that belongs to a later version of Kibana .e.g. `.kibana_7.12.0_001` fail the migration.
+4. Mark the source index as read-only and wait for all in-flight operations to drain (requires new functionality in Elasticsearch). This prevents any further writes from outdated nodes. Assuming this API is similar to the existing `/<index>/_close` API, we expect to receive `"acknowledged" : true` and `"shards_acknowledged" : true`. If all shards don’t acknowledge within the timeout, retry the operation until it succeeds.
+5. Clone the source index into a new target index which has writes enabled. All nodes on the same version will use the same fixed index name e.g. `.kibana_7.10.0_001`. The `001` postfix can be changed with the configuration option i.e. `migrations.target_index_postfix: '002'`
+   1. `POST /.kibana_n/_clone/.kibana_7.10.0_001?wait_for_active_shards=all {"settings": {"index.blocks.write": true}}`. Ignore errors if the clone already exists.
+   2. Wait for the cloning to complete `GET /_cluster/health/.kibana_7.10.0_001?wait_for_status=green&timeout=60s` If cloning doesn’t complete within the 60s timeout, log a warning for visibility and poll again.
+6. Update the mappings of the target index
```

Comment:
Am I right in my understanding that we cannot remove mappings or stop indexing fields with this pattern? In those scenarios, can we still do a regular reindex instead of a clone?

I worry that we may need to remove many mapped fields sooner than later. We are pushing up against the default limit of 1000 indexed fields and I think there are many that do not need to be indexed.

Reply:
> Am I right in my understanding that we cannot remove mappings or stop indexing fields with this pattern? In those scenarios, can we still do a regular reindex instead of a clone?

Yes, we cannot remove mappings. So if a migration "renames" a field we'll be forever left with the old and new fields which will accelerate the field growth. (We can stop indexing nested fields by setting their type to `object` with `enabled: false` but this always applies to all the children of the disabled parent field so it's a blunt hammer.)

`master` currently sets 408 fields so I think we have quite a bit of headroom at least during 7.x There isn't an easy way to track this, we would have to perform an upgrade through all the bug fix releases to come up with a worst-case count so we should probably add telemetry for this at least in 8.0.

The only way to "reset" the mappings is doing a reindex, it will be slower than a clone, but other than the longer downtime window it won't impact the algorithm. We could look into doing a re-index for every major in order to clean up mappings of unused fields.

Longer term we need consider what we'll do to "scale" beyond 1000 fields. We could either "re-balance" objects across multiple indexes or simply up the field limit in the Elasticsearch system indices plugin for Kibana. I don't fully understand the implications of this tradeoff.

Reply:
TIL: we can disable a single field from being indexed with `index: no` https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-index.html

Reply:
Have you been able to change the `index` property of a field from `index: true` to `index: false`? I haven't been able to in my limited testing?

Assuming we aren't able to, that means that any field which was added during any version of Kibana or any plugin of Kibana will be persistent in the `.kibana_${version}` indices whenever the `_clone` is used. Did you have any plans to ensure that for all of 7.x or 8.x that we don't eventually cross this limit, including fields which were left behind by old versions or disabled plugins?

Reply:
Yeah I should have updated this conversation after my further investigation of our field growth. We can't use `index: false` or `enabled: false` to reduce the field count. So when cloning, fields will always increase on every release.

So we basically have a tradeoff between migration performance for huge indices and containing the field growth. As of now, there are no concrete plans for any plugins who want to store > 100k objects or gigabytes of data. So I think as a first step we should rather contain the field growth.

I created https://github.com/elastic/kibana/issues/70471 for scaling field count and want to leave it out of the scope of this RFC. But I think the long term solution to migration performance for huge indices and scaling field count will be the same: use more independent indices. The more granular the indices the higher the ratio of objects requiring a migration vs objects that don't need to be migrated. So on the far side of the spectrum if we go with an index per saved object type then reindexing performance should be nearly equivalent to cloning and transforming performance.

But stable lock-free migrations will be a pre-requisite.

Reply:
I agree that long-term we should be using more independent granular indices. In the short-term, I just wanted to make sure that we didn't accumulate enough abandoned fields over time that we prevent users who frequently upgrade from upgrading without a reindex.

Would it be possible to increase the `index.mapping.total_fields.limit` based on the fields that we know about to make sure that we don't hit this limit until we're able to partition the indices? Is my understanding correct that given the traditional size of the `.kibana*` indices, if we go over the default limit of 1000 fields, the performance penalty isn't expected to be substantial?

Reply:
Yes the ES team's gut feeling is that doubling the field count shouldn't have a noticeable performance effect because of `.kibana` having so few documents.

The other alternative is to use a reindex instead of a clone, it would be a bit slower, but then we can remove fields. The only way for a node to check for completion of a reindex operation started by a different node is to use the tasks API which is in beta and it seems like building this kind of functionality on the tasks API is discouraged https://github.com/elastic/elasticsearch/issues/51628. I've asked the ES team for feedback.

Switching from cloning to reindex should be relatively easy and low risk so it could be done in a future release if we see that the abandoned field count is becoming significant.

Reply:
Just to close the loop here, feedback from the ES team is that we can use the beta tasks API for checking reindex completion as this is the only API exposing this information at the moment. But we'll use this as an escape hatch if we haven't found better alternatives for scaling the field count.